### PR TITLE
refactor: thread PhysicalOptimizerContext through join_selection stats helpers

### LINKED ISSUE #23671: Thread `PhysicalOptimizerContext` through `join_selection` stats helpers (CBO extension pattern)
### Is your feature request related to a problem or challenge?

`JoinSelection`'s internal stats helpers (`get_stats`, `should_swap_join_order`,
`supports_collect_by_thresholds`) take `Option<&StatisticsRegistry>` and
`&ConfigOptions` pre-extracted from the `PhysicalOptimizerContext`. Every new
session-scoped input added to the context for cost-based estimation forces a
signature change across all three helpers and their call sites.

### Describe the solution you'd like

Pass `&dyn PhysicalOptimizerContext` to these helpers instead; they read
`statistics_registry()` / `config_options()` off it. New context inputs then need
no helper-signature changes.

This is also the pattern any future physical optimizer rule implementing CBO
should follow: take the `PhysicalOptimizerContext` and read what it needs off it,
rather than threading individual extracted inputs.

### Describe alternatives you've considered

- Keep threading individual params — every new input churns all three signatures.
- Pass `SessionState` directly — not possible; `datafusion-physical-optimizer`
  can't depend on `datafusion-core`, which is why `PhysicalOptimizerContext`
  exists.

### Additional context

Non-breaking, mechanical: the helpers are crate-internal (`get_stats` /
`supports_collect_by_thresholds` private, `should_swap_join_order` `pub(crate)`),
no cross-crate callers, public `PhysicalOptimizerRule` methods unchanged.

### PR DESCRIPTION
**Pull request by asolimando**
## Which issue does this PR close?

- Closes #23671.

## Rationale for this change

`JoinSelection`'s internal statistics helpers each took separate `config: &ConfigOptions`
and `registry: Option<&StatisticsRegistry>` parameters. Any new session-scoped CBO input
to these rules (e.g. #21120) would require touching every helper's signature.

## What changes are included in this PR?

Replace the `config`/`registry` parameter pair on the internal statistics helpers with a
single `context: &dyn PhysicalOptimizerContext`, reading `context.config_options()` and
`context.statistics_registry()` internally. This is a pure internal signature change: all
affected functions are private or `pub(crate)`, with no cross-crate callers. Behavior is
unchanged.

## Are these changes tested?

No new tests; this is a non-functional refactor covered by existing tests
(`cargo test -p datafusion-physical-optimizer`, the `physical_optimizer::join_selection`
integration tests, and `statistics_registry.slt`).

## Are there any user-facing changes?

No.

----
Disclaimer: I used AI to assist in the code generation, I have manually reviewed the output and it matches my intention and understanding.

### CODE REVIEW

**Comment by asolimando** [comment]
@kosiew a small follow-up of https://github.com/apache/datafusion/pull/23651, this is a pure refactor to properly propagate `PhysicalOptimizerContext`, which is meant to be a convenient way to pass context to physical rules, especially those dealing with CBO, so we can enrich context for statistics propagation without breaking signatures.

I have a question: since it's a pure refactor and it doesn't really alter coverage, I would consider https://github.com/apache/datafusion/pull/24716#issuecomment-5436734704, wdyt?

Thanks!

File: datafusion/physical-optimizer/src/join_selection.rs [hunk 1]
```diff
@@ -152,19 +151,15 @@ impl PhysicalOptimizerRule for JoinSelection {
...
         ];
         let new_plan = plan
-            .transform_up(|p| apply_subrules(p, &subrules, config))
+            .transform_up(|p| apply_subrules(p, &subrules, context.config_options()))
             .data()?;
         new_plan
-            .transform_up(|plan| {
-                statistical_join_selection_subrule(plan, config, registry)
-            })
+            .transform_up(|plan| statistical_join_selection_subrule(plan, context))
```
Comment (kosiew) [inline datafusion/physical-optimizer/src/join_selection.rs:162]: Could we add a focused test that calls `JoinSelection::optimize_with_context` with a custom context and statistics registry where the estimates actually change the join decision? The existing registry SLT covers this through the `SessionState` path, but a test here would directly exercise the context-threading boundary introduced by this refactor. Not blocking, but I think it would be useful coverage.

**Review by kosiew — APPROVED** [review]
@asolimando, thanks for working on this. The refactor looks good to me. I don't see any blocking issues. I left one non-blocking suggestion around adding focused coverage for the new context-threading boundary.

File: datafusion/physical-optimizer/src/join_selection.rs (hunk 1 above)
Comment (asolimando) [reply datafusion/physical-optimizer/src/join_selection.rs:162]: Thanks for your review! It's indeed a good chance to improve coverage on `JoinSelection::optimize_with_context` itself, I have proposed a test along the line you proposed in https://github.com/apache/datafusion/pull/24716/commits/809b30b192992745b03fe4a82a441c70780a5c6d, let me know if that works for you.

**Comment by kosiew** [comment]
🚀 
@asolimando 
Thank you for your contribution.
